Rubberduck.Inspections/Concrete/ShadowedDeclarationInspection.cs

```diff
@@ -0,0 +1,628 @@
+using Rubberduck.Inspections.Results;
+using Rubberduck.Parsing;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Inspections.Abstract;
+using Rubberduck.Parsing.Inspections.Resources;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.UI;
+using Rubberduck.VBEditor.SafeComWrappers;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    public sealed class ShadowedDeclarationInspection : InspectionBase
+    {
+        private enum DeclarationSite
+        {
+            NotApplicable = 0,
+            ReferencedProject = 1,
+            OtherComponent = 2,
+            SameComponent = 3
+        }
+
+        private class OptionPrivateModuleListener : VBAParserBaseListener
+        {
+            public List<VBAParser.ModuleContext> OptionPrivateModules { get; } = new List<VBAParser.ModuleContext>();
+
+            public override void EnterModule(VBAParser.ModuleContext context)
+            {
+                if (context.FindChildren<VBAParser.OptionPrivateModuleStmtContext>().Any())
+                {
+                    OptionPrivateModules.Add(context);
+                }
+            }
+        }
+
+        private class EnumerationRuleIndexListener : VBAParserBaseListener
+        {
+            public Dictionary<ParserRuleContext, int> DeclarationIndexes { get; } = new Dictionary<ParserRuleContext, int>();
+
+            public override void EnterEnumerationStmt(VBAParser.EnumerationStmtContext context)
+            {
+                DeclarationIndexes.Add(context, context.Start.TokenIndex);
+            }
+
+            public override void EnterEnumerationStmt_Constant(VBAParser.EnumerationStmt_ConstantContext context)
+            {
+                DeclarationIndexes.Add(context, context.Start.TokenIndex);
+            }
+        }
+
+        public ShadowedDeclarationInspection(RubberduckParserState state) : base(state)
+        {
+        }
+
+        public override Type Type => typeof(ShadowedDeclarationInspection);
+
+        public override CodeInspectionType InspectionType { get; } = CodeInspectionType.CodeQualityIssues;
+
+        public override IEnumerable<IInspectionResult> GetInspectionResults()
+        {
+            var optionPrivateModuleListener = new OptionPrivateModuleListener();
+            var enumerationRuleIndexListener = new EnumerationRuleIndexListener();
+
+            foreach (var module in State.DeclarationFinder.AllModules.Where(m => m.ComponentType == ComponentType.StandardModule))
+            {
+                ParseTreeWalker.Default.Walk(optionPrivateModuleListener, State.GetParseTree(module));
+            }
+
+            foreach (var module in State.AllUserDeclarations.Where(d => d.DeclarationType == DeclarationType.ProceduralModule || d.DeclarationType == DeclarationType.ClassModule))
+            {
+                ParseTreeWalker.Default.Walk(enumerationRuleIndexListener, State.GetParseTree(module.QualifiedName.QualifiedModuleName));
```

Comment:
With visitors, you would call `visitor.Visit()` to collect the results. instead of walking the parse tree.